### PR TITLE
esp-idf: Add more socket support for esp-idf

### PR DESCRIPTION
**Pull request by donghengqaz**
esp-idf: Add socket support for esp-idf platform
    
1. Add OS socket functions
2. Support to get/set recv_buf_size & send_buf_size & reuse_port & reuse_addr

### CODE REVIEW

**Comment by wenyongh** [comment]
@donghengqaz There are link errors in SGX CI:
https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/3318646884/jobs/5482961112
```
/usr/bin/ld: libvmlib.a(posix.c.o): in function `wasi_ssp_sock_set_recv_buf_size':
posix.c:(.text.wasi_ssp_sock_set_recv_buf_size+0x89): undefined reference to `setsockopt'
/usr/bin/ld: libvmlib.a(posix.c.o): in function `wasi_ssp_sock_set_reuse_addr':
posix.c:(.text.wasi_ssp_sock_set_reuse_addr+0x8c): undefined reference to `setsockopt'
/usr/bin/ld: libvmlib.a(posix.c.o): in function `wasi_ssp_sock_set_reuse_port':
posix.c:(.text.wasi_ssp_sock_set_reuse_port+0x8c): undefined reference to `setsockopt'
/usr/bin/ld: libvmlib.a(posix.c.o): in function `wasi_ssp_sock_set_send_buf_size':
posix.c:(.text.wasi_ssp_sock_set_send_buf_size+0x89): undefined reference to `setsockopt'
```
Could you add empty implementation of setsockopt in SGX sgx_socket.c and sgx_socket.h, similar to getsockopt?
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/shared/platform/linux-sgx/sgx_socket.h
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/shared/platform/linux-sgx/sgx_socket.c#L354-L371

And refer to:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/shared/platform/linux-sgx/sgx_socket.c#L910-L916

**Comment by wenyongh** [comment]
@loganek Could you help review this PR? Thanks.

File: core/shared/platform/linux-sgx/sgx_socket.c [hunk 1]
```diff
@@ -370,6 +370,24 @@ getsockopt(int sockfd, int level, int optname, void *optval, socklen_t *optlen)
     return ret;
 }
 
+int
+setsockopt(int sockfd, int level, int optname, const void *optval, socklen_t optlen);
```
Comment (wenyongh) [inline core/shared/platform/linux-sgx/sgx_socket.c:374]: Should remove `;`

File: core/shared/platform/esp-idf/espidf_socket.c [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+{
+    assert(textual);
+
+    out->sin_family = AF_INET;
+    out->sin_port = htons(port);
+    out->sin_addr.s_addr = inet_addr(textual);
+}
+
+int
+os_socket_create(bh_socket_t *sock, int tcp_or_udp)
```
Comment (wenyongh) [inline core/shared/platform/esp-idf/espidf_socket.c:22]: Should be `os_socket_create(bh_socket_t *sock, bool is_tcp)`, see definition in platform_api_extension.h

File: core/shared/platform/esp-idf/espidf_socket.c [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+{
+    if (!sock) {
+        return BHT_ERROR;
+    }
+
+    if (1 == tcp_or_udp) {
```
Comment (wenyongh) [inline core/shared/platform/esp-idf/espidf_socket.c:28]: Should be `if (is_tcp)`

File: core/shared/platform/esp-idf/espidf_socket.c [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        *sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+    }
+    else if (0 == tcp_or_udp) {
```
Comment (wenyongh) [inline core/shared/platform/esp-idf/espidf_socket.c:31]: Should be `else {`

File: core/shared/platform/esp-idf/espidf_socket.c [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    if (ret < 0) {
+        goto fail;
+    }
+
+    socklen = sizeof(addr);
+    if (getsockname(socket, (void *)&addr, &socklen) == -1) {
+        goto fail;
+    }
+
+    *port = ntohs(addr.sin_port);
```
Comment (wenyongh) [inline core/shared/platform/esp-idf/espidf_socket.c:62]: The implementation of posix is a little more complex:
```C
    ret = fcntl(socket, F_SETFD, FD_CLOEXEC);
    if (ret < 0) {
        goto fail;
    }

    ret = setsockopt(socket, SOL_SOCKET, SO_LINGER, &ling, sizeof(ling));
    if (ret < 0) {
        goto fail;
    }

    if (!textual_addr_to_sockaddr(host, *port, (struct sockaddr *)&addr,
                                  &socklen)) {
        goto fail;
    }

    ret = bind(socket, (struct sockaddr *)&addr, socklen);
    if (ret < 0) {
        goto fail;
    }

    socklen = sizeof(addr);
    if (getsockname(socket, (void *)&addr, &socklen) == -1)
```
Could you check whether it is OK here?

Comment (donghengqaz) [reply core/shared/platform/esp-idf/espidf_socket.c:62]: At fisrt, we copy the related code from other platform and test it, but it fails in esp-idf platform, so we modify code to adjust  esp-idf's LwIP.

1. 
```
    ret = fcntl(socket, F_SETFD, FD_CLOEXEC);
    if (ret < 0) {
        goto fail;
    }
```
socket of esp-idf's LwIP doesn't support `F_SETFD` and `FD_CLOEXEC`.

2.
```
ret = setsockopt(socket, SOL_SOCKET, SO_LINGER, &ling, sizeof(ling));
    if (ret < 0) {
        goto fail;
    }
```
TCP socket of esp-idf's LwIP doesn't use this option by default, so we remove these process.

Comment (wenyongh) [reply core/shared/platform/esp-idf/espidf_socket.c:62]: OK, got it, thanks a lot.

File: core/shared/platform/include/platform_api_extension.h [hunk 6]
```diff
@@ -977,6 +977,36 @@ os_socket_set_broadcast(bh_socket_t socket, bool is_enabled);
 int
 os_socket_get_broadcast(bh_socket_t socket, bool *is_enabled);
 
+/**
+ * Get the peer address of the connected to the socket
+ *
+ * @param socket  [INTPUT] the connected socket
+ * @param addr    [OUTPUT] the address of the peer socket. If addr is NULL,
+ *                nothing is filled in, and addrlen will not be used
```
Comment (wenyongh) [inline core/shared/platform/include/platform_api_extension.h:985]: No addrlen argument here? The prototype is different from libc function prototype, and the implementation of esp-idf platform sets the input size as `socklen_t addr_len = sizeof(struct sockaddr_in);`, so could you please add comment about it (like the size of addr should be no smaller than sizeof(struct sockaddr_int)?

File: core/shared/platform/include/platform_api_extension.h [hunk 7]
```diff
@@ -977,6 +977,36 @@ os_socket_set_broadcast(bh_socket_t socket, bool is_enabled);
...
+ * @param port    [OUTPUT] the port number, if the value is 0,
+ *                it will use a port assigned by OS. On return it will
+ *                contain the actual bound port number
+ *
+ * @return 0 if success, -1 otherwise
+ */
+int
+os_socket_getpeername(bh_socket_t socket, uint32 *addr, int *port);
```
Comment (wenyongh) [inline core/shared/platform/include/platform_api_extension.h:993]: Had better use `struct sockaddr *addr`, keep it same as libc argument type.

File: core/shared/platform/include/platform_api_extension.h [hunk 8]
```diff
@@ -977,6 +977,36 @@ os_socket_set_broadcast(bh_socket_t socket, bool is_enabled);
...
+
+/**
+ * Get the current address of the bound to the socket
+ *
+ * @param socket  [INTPUT] the bound socket
+ * @param addr    [OUTPUT] the address of the bound socket. If addr is NULL,
+ *                nothing is filled in, and addrlen will not be used
```
Comment (wenyongh) [inline core/shared/platform/include/platform_api_extension.h:1000]: Same as above

File: core/shared/platform/include/platform_api_extension.h [hunk 9]
```diff
@@ -977,6 +977,36 @@ os_socket_set_broadcast(bh_socket_t socket, bool is_enabled);
...
+ * @param port    [OUTPUT] the port number, if the value is 0,
+ *                it will use a port assigned by OS. On return it will
+ *                contain the actual bound port number
+ *
+ * @return 0 if success, -1 otherwise
+ */
+int
+os_socket_getsockname(bh_socket_t socket, uint32 *addr, int *port);
```
Comment (wenyongh) [inline core/shared/platform/include/platform_api_extension.h:1008]: Same as above

**Review by wenyongh — COMMENTED** [review]
LGTM

**Comment by donghengqaz** [comment]
Thanks @wenyongh, some APIs were changed in recent WASM version.

**Pull request merged**